### PR TITLE
Add: scheduled query backoff in case of errors

### PR DESCRIPTION
**Pull request by washort**
This tracks scheduled query model objects all through the celery task for execution so that failure can be recorded. Number of failures are then used to decide how much longer to delay query execution.

### CODE REVIEW

File: redash/tasks/queries.py [hunk 1]
```diff
@@ -441,9 +457,13 @@ def run(self):
 
     def _annotate_query(self, query_runner):
         if query_runner.annotate_query():
+            if self.task.request.delivery_info:
```
Comment (arikfr) [inline redash/tasks/queries.py]: Why is this needed?

File: migrations/versions/d1eae8b9893e_.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+# revision identifiers, used by Alembic.
+revision = 'd1eae8b9893e'
+down_revision = '65fc9ede4746'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.add_column('queries', sa.Column('schedule_failures', sa.Integer(),
```
Comment (arikfr) [inline migrations/versions/d1eae8b9893e_.py]: I wonder if we should store this in the database. Maybe in Redis instead?

The main benefit of keeping it in the database is that we can easily query for it, to generate a report of failing queries (although then maybe we need `last_failure_at` column too?).

The downside is that it makes the table bloated with operational data and adds more calls to the database.

File: redash/tasks/queries.py [hunk 3]
```diff
@@ -425,7 +435,13 @@ def run(self):
         if error:
             self.tracker.update(state='failed')
             result = QueryExecutionError(error)
+            if self.scheduled_query:
+                self.scheduled_query.schedule_failures += 1
+                models.db.session.add(self.scheduled_query)
         else:
+            if self.scheduled_query:
+                self.scheduled_query.schedule_failures = 0
```
Comment (arikfr) [inline redash/tasks/queries.py]: How about we do this update only if `schedule_failures` is `> 0`? Otherwise we generate unneeded update.

File: redash/tasks/queries.py (hunk 1 above)
Comment (washort) [inline redash/tasks/queries.py]: this is None if invoked outside of celery (such as in unit tests)

File: migrations/versions/d1eae8b9893e_.py (hunk 2 above)
Comment (washort) [inline migrations/versions/d1eae8b9893e_.py]: I wondered about using Redis but wasn't clear on how it should be done. I'll have another crack at it if that's what you'd prefer.

File: redash/tasks/queries.py (hunk 1 above)
Comment (arikfr) [inline redash/tasks/queries.py]: :\ I don't like changing code to accommodate for tests (this will never be executed outside of a celery context otherwise). Two alternatives:

- Pass the routing key & task id to the constructor instead of the task directly. 
- Mock this.

WDYT?

File: migrations/versions/d1eae8b9893e_.py (hunk 2 above)
Comment (arikfr) [inline migrations/versions/d1eae8b9893e_.py]: Both options have their benefits. Using Redis will allow us to store more metadata without feeling we're abusing the queries table :) I mean metadata like:

- What the error was.
- When the last error occurred.

**Comment by washort** [comment]
Fixed the things you mentioned; mocking the celery stuff for test turned out to be easier than I thought.

**Comment by arikfr** [comment]
Just realized something: we need to reset the `schedule_failures` counter when the user changes the query text.

**Comment by washort** [comment]
Good point! Added.

**Comment by arikfr** [comment]
Thanks!
